Bus monitor polish

scothis opened the pull request:
General bus enhancements lifted from #378, since that PR is blocked.

- avoid unnessesary status updates if the status has not changed
- avoid supressing errors when updating status
- avoid supressing updates for unchanged resources as the local copy may not be fully synched
- addressed all govet and golint reports from pkg/buses
 
/assign @n3wscott

n3wscott commented:
/lgtm

Minor comments mostly not related to this change but to the code in general.

evankanderson commented on pkg/buses/kafka/dispatcher/dispatcher.go:
```diff
@@ -187,14 +187,14 @@ func (d *dispatcher) subscribe(subscription *channelsv1alpha1.Subscription, para
-	sInitial := parameters[InitialOffset]
-	if sInitial == Oldest {
+func resolveInitialOffset(parameters buses.ResolvedParameters) (int64, error) {
+	sInitial := parameters[initialOffset]
```
Switch or a map would be slightly cleaner here; i.e.
```golang
const saramaOffset = map[string]int64{
    "newest": saramaNewest,
    "oldest": saramaOldest,
}

func resolveInitialOffset(...) (int64, error) {
  offset, ok := saramaOffset[parameters[initialOffset]
  if !ok {
    return 0, fmt.Errorf(...)
  }
  return offset, nil
}
```

evankanderson commented on pkg/buses/monitor.go:
```diff
@@ -192,9 +203,16 @@ func (h MonitorEventHandlerFuncs) onSubscribe(subscription *channelsv1alpha1.Sub
-			glog.Warningf("Could not update status: %v", errS)
+		if !equality.Semantic.DeepEqual(subscription.Status, subscriptionCopy.Status) {
+			// status has changed, update subscription
+			_, errS := monitor.eventingClient.ChannelsV1alpha1().Subscriptions(subscription.Namespace).Update(subscriptionCopy)
```
It feels strange to be tracking two `error`s here.

scothis replied:
it is a bit funky. We want always want to update the resource status, if the original err being recorded in the resource status, we want to return that error since it's more meaningful. If the update failed, we still want to return that error to trigger the work queue to process the resource again as the update can fail because of a failed optimistic lock.

Will add a comment in the code

evankanderson commented on pkg/buses/references.go:
```diff
@@ -16,22 +16,86 @@
+
 func (r *BusReference) String() string {
-	return fmt.Sprintf("%s/%s", r.Namespace, r.Name)
+	if len(r.Namespace) != 0 {
```
I have a slight preference for `r.Namespace != ""`; it makes bugs like #377 slightly easier to catch, IMO.

scothis replied:
done

scothis commented:
@n3wscott @evankanderson PTAL

I addressed your feedback and also pulled in the other refactoring I've been working on. The diff is rather scary, but mostly is moving items around to create components that are similar, more cleanly defined and testable. There's more to do (especially adding tests and polishing rough spots), but I like the general direction and it's a better foundation.

From the commit message:

```
Clarify and solidify the public interfaces for a Bus. Common behavior
has moved into composable units that can individually tests or injected
into the bus for testing.

Bus implementors generally only need to provide custom implementations
for up to six methods defined by EventHandlerFuncs.

Core components:
- [Bus|Channel|Subscription]Reference: decouples bus from the raw
    k8s resources
- Cache: lookup and presist raw resource types for a reference
- EventHandlerFuncs: the contract for bus implementations
- Reconciler: reads resource changes from the API server, saves the
    resource in the Cache and emits events to the EventHandlerFuncs
- MessageDispatcher: sends messages to a destination over the event
    delivery protocol
- MessageReceiver: receive messages for a channel on the bus over the
    event delivery protocol

Behavior changes:
- avoid unnessesary status updates if the status has not changed
- avoid supressing errors when updating status
- avoid supressing updates for unchanged resources as the local copy may
  not be fully reconciled
```

scothis commented on pkg/buses/bus.go:
```diff
@@ -0,0 +1,166 @@
+// EventHandlerFuncs are used to be notified when a subscription is created,
+// updated or removed, or a message is received.
+func NewBusDispatcher(busRef BusReference, handlerFuncs EventHandlerFuncs, opts *BusOpts) BusDispatcher {
+	bus := &bus{
```
will do, although it's not quite as clean as you'd sugest because there's a cyclic dependency between the bus and the receiver I still need to untangle.

scothis commented on pkg/buses/bus.go:
```diff
@@ -0,0 +1,166 @@
+		handlerFuncs: handlerFuncs,
+	}
+
+	if opts.Cache == nil {
```
👍

scothis commented on pkg/buses/bus.go:
```diff
@@ -0,0 +1,166 @@
+}
+
+func (b *bus) dispatchMessage(subscription *channelsv1alpha1.Subscription, message *Message) error {
+	subscriber := subscription.Spec.Subscriber
```
I can't think of a reason to why it should be allowed to be empty. I'll add a check to the validator.

scothis commented on pkg/buses/gcppubsub/bus.go:
```diff
@@ -166,93 +127,158 @@ func (b *PubSubBus) ReceiveEvents(sub *channelsv1alpha1.Subscription, parameters
+		return nil
+	}
+
+	glog.Infof("Delete topic %q\n", topicID)
```
yes, but in a follow-up PR

scothis commented on pkg/buses/gcppubsub/bus.go:
```diff
@@ -166,93 +127,158 @@ func (b *PubSubBus) ReceiveEvents(sub *channelsv1alpha1.Subscription, parameters
+		return err
+	} else if exists {
+		// TODO update subscription configuration
+		// _, err := subscription.Update(b.ctx, pubsub.SubscriptionConfigToUpdate{})
```
It's not important right now because there are no bus params, thus the pub/sub subscription will never need to be mutated.

scothis replied:
amending the current comment to make it more clear

scothis commented on pkg/buses/gcppubsub/bus.go:
```diff
@@ -19,139 +19,100 @@ package gcppubsub
+// Knative Subscription. This method will not block, but will continue to
+// receive events until either this method or StopReceivingEvents is called for
+// the same Subscription.
+func (b *CloudPubSubBus) startReceivingEvents(subscriptionRef buses.SubscriptionReference, parameters buses.ResolvedParameters) error {
```
I'll submit a follow-up PR

scothis commented on pkg/buses/handler_funcs.go:
```diff
@@ -0,0 +1,258 @@
+		err = h.ProvisionFunc(channelRef, parameters)
+		channelCopy := channel.DeepCopy()
+		var cond *channelsv1alpha1.ChannelCondition
+		if err != nil {
```
updating

scothis commented on pkg/buses/handler_funcs.go:
```diff
@@ -0,0 +1,258 @@
+			_, errS := reconciler.eventingClient.ChannelsV1alpha1().Channels(channel.Namespace).Update(channelCopy)
+			if errS != nil {
+				glog.Warningf("Could not update channel status: %v", errS)
+				if err != nil {
```
will smoosh them into a single error

scothis commented on pkg/buses/handler_funcs.go:
```diff
@@ -0,0 +1,258 @@
+func (h EventHandlerFuncs) onUnprovision(channel *channelsv1alpha1.Channel, reconciler *Reconciler) error {
+	if h.UnprovisionFunc != nil {
+		channelRef := NewChannelReference(channel)
+		err := h.UnprovisionFunc(channelRef)
```
updating

scothis commented:
@n3wscott PTAL

n3wscott commented:
/lgtm

I think this is a really nice refactoring.

scothis commented:
for approval:
/assign @evankanderson

evankanderson commented on pkg/buses/handler_funcs.go:
```diff
@@ -0,0 +1,253 @@
+}
+
+func (h EventHandlerFuncs) onSubscribe(subscription *channelsv1alpha1.Subscription, reconciler *Reconciler) error {
+	if h.SubscribeFunc != nil {
```
In the future, convert this to early-return:

```golang
if h.SubscribeFunc == nil {
 return nil
}
```

scothis commented:
@evankanderson PTAL. I resolved the conflict with master and addressed your comment
